### DISCUSSION TITLE
Running excalibur on the server

### DISCUSSION THREAD

**Discussion by savaparoski** [excaliburjs/Excalibur#2759] [Ideas]
There was a [short conversation](https://github.com/excaliburjs/Excalibur/discussions/2745) about running excalibur on the server. 
I looked at the example @eonarheim provided and it works just fine.

However, this means I have to replicate a lot of my code into matter.js and then make sure it works just like in excalibur in order to simulate game on the server.

Are there any ways to run excalibur on the server from the same codebase as clients?

**Comment by eonarheim** [comment]
@savaparoski Let me dig around, it should be possible to run most systems (apart from graphics) in a server context. I'll publish a sample when I figure out something that works.

Totally agree that re-using the existing excalibur code would be ideal!

**Reply by eonarheim** [reply]
So at the moment, the latest version of excalibur resists running on node due to the webpack build configuration, I'm working through the build to see if I can't get something working locally 🤞

**Reply by savaparoski** [reply]
Thanks for the effort @eonarheim 

Did you make any progress?

**Reply by eonarheim** [reply]
Hi @savaparoski!

I've made some progress, currently it only works on a local version of excalibur using `npm link` but I should be able to cut a patch soon.

Example repo: https://github.com/excaliburjs/sample-physics-server
Excalibur branch: https://github.com/excaliburjs/Excalibur/tree/bugfix/run-excalibur-in-node

Still a work in progress, I think the simulation could be a lot smoother on the frontend with some linear interpolation between server updates

**Comment by singerbj** [comment]
Looks like this thread is based around running excaliburjs on the client and a physics runtime on the server-side....what about running excaliburjs in a "headless" mode on the server? Or even the use of an appropriate shim on top of the main library that allows for this?

My goal would be to run a authoritative instance of my multiplayer game on a server, using the exact same codebase as the client.

**Reply by eonarheim** [reply]
@singerbj Totally agree on the desire to run an authoritative instance, I'm working on a "headless" mode in that branch.

Currently I'm trying something like this

```typescript
const serverGame = new ex.Engine({headless: true});

setInterval(() => {

  // tick the mainloop
  serverGame.mainloop(....)
}, 60);
```

The tricky part is the engine is currently fairly coupled to the DOM and graphics contexts, so splitting those has proved to be a challenge.

**Reply by singerbj** [reply]
Ah ok I misunderstood what you had written above. Happy to help out if I can - might have some time this weekend to more deeply review the inner workings of excaliburjs. If you have any specific problems that need more eyes, let me know.

**Reply by singerbj** [reply]
Well won't be able to look much further now this weekend, but maybe this could help to improve on the Polyfill you started to accommodate a nodejs runtime? https://www.npmjs.com/package/mock-browser

**Reply by eonarheim** [reply]
Thanks @singerbj I hadn't considered that angle, I'll take a look!

**Reply by singerbj** [reply]
Sure thing. Know I did I try to take a look at the inner workings of Excalibur but things appear very tightly coupled to the DOM (which makes sense) and so I got a little overwhelmed and now am busy.

High level though accomplishing this see like it could be done with an expanded polyfill or a rewrite to abstract DOM dependant code from non-DOM dependent code. I bet you'd lean toward the former :)

Also seems like you could put a lot on the "user" here, and have them manage when to use certain graphical aspects of the library..like "don't use sprites on your server". Basically the you'd only have to do the minimum on the library side to ensure core aspects of it can run server side.

**Reply by singerbj** [reply]
Any updates on this?

I did tinker a bit but there didn't seem like any really easy solution...my latest idea was to consider using headless electron as a way to have this work. It's going decent so far but I don't think that its the ideal solution long term.

**Reply by eonarheim** [reply]
@singerbj Apologies, no major progress on my front. This will take some refactoring to accomplish, but it is still something I want to support long term

**Reply by singerbj** [reply]
No worries @eonarheim! Glad to hear it's in the plans. Naturally I took a stab at it again and quickly got overwhelmed :). Wish there was a good drop in solution for this. Definitely willing to help out when the time comes...if I have time!

**Reply by singerbj** [reply]
Ok @eonarheim so my addictive personality could not let this rest - I think I may have gotten something at least somewhat feasible using mocks/shims. Check this out: https://github.com/singerbj/excalibur-sample-in-node/tree/main

If you clone, just ```npm i``` and ```npm run start``` and you should see console output for the bot in the sample moving back and forth. (I was using Node.js v18.7.1 for whatever reason.)

For code, take a look at 'src/browserShim.ts' that allows for this to work. It's a massive hack but kinda pumped I got it to work. I based it off of findings I made for some of the other js game engines like Phaser and Pixi.

Note that I did not attempt to make the same codebase work on the browser at the same time.

**Reply by singerbj** [reply]
I'm still not sure this is exactly the official approach you would want to take to make this work. The purist in me says that you would want to refactor excalibur to have a mode that works as efficiently as possible in a non-browser env - but maybe this works? Especially if it does work for Phaser and Pixi? Maybe the overall use case here is not really even worth it?

**Reply by eonarheim** [reply]
@singerbj Many thanks for digging so hard into this! Honestly this is way easier than the refactoring, I might just say this is the way ❤️

**Reply by eonarheim** [reply]
Perhaps we add an engine flag that drops a shim into the global context? or maybe just a global export? That would make it theoretically easy to support both server and web, you'd need to have a factory function that outs the right engine or runs the shim?

**Reply by singerbj** [reply]
I'll continue to poke around with it - I am sure a more fully fledged project might need more finessing. I'll keep you updated!

**Reply by singerbj** [reply]
> Perhaps we add an engine flag that drops a shim into the global context? or maybe just a global export? That would make it theoretically easy to support both server and web, you'd need to have a factory function that outs the right engine or runs the shim?

Oh just saw this - could do that, or could offer the shim as a separate package that you just need to run before Excalibur?

See how this package works: https://github.com/geckosio/phaser-on-nodejs

**Comment by KostarSf** [comment]
I'm also really looking forward to Excalibur on the server! Right now, I have implemented multiplayer in my game using some kind of distributed computing 😀 Each player locally processes himself and some of the background objects. The server only forwards events between players via web sockets. It's not a very secure system, but it works

**Reply by eonarheim** [reply]
Me too! It's part of the plan before v1, it might make the v0.30 cut but no promises :)

**Reply by KostarSf** [reply]
Can't wait for this!
